### PR TITLE
upgrade to latest typescript

### LINKED ISSUE #2682: Typescript 3.x drop support for WebKit* Types
The File-tree-model is using WebKitEntry, WebKitFileEntry and WebKitDirectoryEntry eg https://github.com/theia-ide/theia/blob/d656a7abfc42346a4080e7f8665a7b42bf6fc7db/packages/filesystem/src/browser/file-tree/file-tree-model.ts#L181.
This is not compatible with Typescript 3.x as they removed WebKit* types see https://github.com/Microsoft/TypeScript/issues/25515#issuecomment-403897228
So instead of WebKitEntry etc FileSystemEntry etc from the File API should be used (https://developer.mozilla.org/en-US/docs/Web/API/FileSystemEntry) .

### LINKED ISSUE #3262: [Question] TS compiler errors in @theia/workspace package
I have an npm package which depends on `@theia/languages` which pulls in `@theia/workspace` dependency.
When I attempt to TS compile my package (with TS 3.1.3, I've tried 2.8.x) i get these errors:

```
ode_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:48:45 - error TS2304: Cannot find name 'WebKitEntry'.

48     protected uploadEntry(base: URI, entry: WebKitEntry | null): void;
                                               ~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:49:54 - error TS2304: Cannot find name 'WebKitDirectoryEntry'.

49     protected uploadDirectoryEntry(base: URI, entry: WebKitDirectoryEntry): Promise<void>;
                                                        ~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:54:34 - error TS2304: Cannot find name 'WebKitDirectoryEntry'.

54     protected readEntries(entry: WebKitDirectoryEntry, cb: (items: any) => void): void;
                                    ~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:55:49 - error TS2304: Cannot find name 'WebKitEntry'.

55     protected uploadEntries(base: URI, entries: WebKitEntry[]): void;
                                                   ~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:56:49 - error TS2304: Cannot find name 'WebKitFileEntry'.

56     protected uploadFileEntry(base: URI, entry: WebKitFileEntry): void;
                                                   ~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:135:28 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

135     const validateSchema = new Ajv().compile({
                               ~~~~~~~~~~~~~~~~~~~
136         type: 'object',
    ~~~~~~~~~~~~~~~~~~~~~~~
... 
151         }
    ~~~~~~~~~
152     });
    ~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:135:32 - error TS2304: Cannot find name 'Ajv'.

135     const validateSchema = new Ajv().compile({
                                   ~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:153:58 - error TS1183: An implementation cannot be declared in ambient contexts.

153     export function is(data: any): data is WorkspaceData {
                                                             ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:156:74 - error TS1183: An implementation cannot be declared in ambient contexts.

156     export function buildWorkspaceData(folders: string[]): WorkspaceData {
                                                                             ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:161:103 - error TS1183: An implementation cannot be declared in ambient contexts.

161     export function transformToRelative(data: WorkspaceData, workspaceFile?: FileStat): WorkspaceData {
                                                                                                          ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:162:38 - error TS1039: Initializers are not allowed in ambient contexts.

162         const folderUris: string[] = [];
                                         ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:163:34 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

163         const workspaceFileUri = new URI(workspaceFile ? workspaceFile.uri : '').withScheme('file');
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:165:31 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

165             const folderUri = new URI(path).withScheme('file');
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:166:25 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

166             const rel = workspaceFileUri.parent.relative(folderUri);
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:167:13 - error TS1036: Statements are not allowed in ambient contexts.

167             if (rel) {
                ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:168:17 - error TS1036: Statements are not allowed in ambient contexts.

168                 folderUris.push(rel.toString());
                    ~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:171:17 - error TS1036: Statements are not allowed in ambient contexts.

171                 folderUris.push(folderUri.toString());
                    ~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:176:103 - error TS1183: An implementation cannot be declared in ambient contexts.

176     export function transformToAbsolute(data: WorkspaceData, workspaceFile?: FileStat): WorkspaceData {
                                                                                                          ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:178:39 - error TS1039: Initializers are not allowed in ambient contexts.

178             const folders: string[] = [];
                                          ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:179:13 - error TS1036: Statements are not allowed in ambient contexts.

179             for (const folder of data.folders) {
                ~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:180:30 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

180                 const path = folder.path;
                                 ~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:181:17 - error TS1036: Statements are not allowed in ambient contexts.

181                 if (path.startsWith('file:///')) {
                    ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:182:21 - error TS1036: Statements are not allowed in ambient contexts.

182                     folders.push(path);
                        ~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:185:21 - error TS1036: Statements are not allowed in ambient contexts.

185                     folders.push(new URI(workspaceFile.uri).withScheme('file').parent.resolve(path).toString());
                        ~~~~~~~

```

They are coming mostly from `@theia/workspace` and its dependencies.

Any idea how to fix or workaround that?

### PR DESCRIPTION
**Pull request by akosyakov**
Closes #2682, closes #3262

### CODE REVIEW

**Comment by kittaakos** [comment]
I am reviewing it now...

File: packages/core/src/browser/tree/search-box.ts [hunk 1]
```diff
@@ -172,7 +172,7 @@ export class SearchBox extends BaseWidget {
 
         const input = document.createElement('input');
         input.readOnly = true;
-        input.onselectstart = () => false;
+        input.addEventListener('selectstart', () => false);
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/core/src/browser/tree/search-box.ts]: ok, I will move it to `onAfterAttach` widget callback that it gets removed on detached callback

File: packages/filesystem/src/typings/dom.webkit.d.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+type WebKitEntriesCallback = ((entries: WebKitEntry[]) => void) | { handleEvent(entries: WebKitEntry[]): void; };
```
Comment (AlexTugarev) [inline packages/filesystem/src/typings/dom.webkit.d.ts]: @akosyakov, have you copied the typings from anywhere? if so, should we reference the source?

Comment (akosyakov) [reply packages/filesystem/src/typings/dom.webkit.d.ts]: not really, I've typed what is necessary to quiet tsc

Comment (AlexTugarev) [reply packages/filesystem/src/typings/dom.webkit.d.ts]: thanks!

File: packages/core/src/browser/tree/search-box.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/core/src/browser/tree/search-box.ts]: updated, if the build is green good to merge?

[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/core/src/browser/tree/search-box.ts]: 👍

**Pull request merged**